torch/lowrank.py

```diff
@@ -0,0 +1,325 @@
+"""Implement various linear algebra algorithms for low rank matrices.
+"""
+
+import torch
+
+
+def is_sparse(A):
+    """Check if tensor A is a sparse tensor"""
+    if isinstance(A, torch.Tensor):
+        return A.layout == torch.sparse_coo
+    import scipy.sparse
+    return isinstance(A, scipy.sparse.base.spmatrix)
+
+
+def uniform(low=0.0, high=1.0, size=None, dtype=None, device=None):
+    """Returns a tensor filled with random numbers from a uniform
+    distribution on the interval :math:`[low, high)`
+
+    Arguments::
+
+      size (int...): a sequence of integers defining the shape of the
+        output tensor.  Can be a variable number of arguments or a
+        collection like a list or tuple.
+
+      dtype (:class:`torch.dtype`, optional): the desired data type of
+        returned tensor.  Default: if ``None``, uses a global default
+        (see :func:`torch.set_default_tensor_type`).
+
+      device (:class:`torch.device`, optional): the desired device of
+        returned tensor.  Default: if ``None``, uses the current
+        device for the default tensor type (see
+        :func:`torch.set_default_tensor_type`). :attr:`device` will be
+        the CPU for CPU tensor types and the current CUDA device for
+        CUDA tensor types.
+
+    """
+    attrs = dict(dtype=dtype, device=device)
+    if size is None:
+        r = low + (high - low) * torch.rand(1, **attrs)[0]
+    else:
+        r = low + (high - low) * torch.rand(*size, **attrs)
+    if dtype in [torch.complex32, torch.complex64, torch.complex128]:
+        if size is None:
+            i = low + (high - low) * torch.rand(1, **attrs)[0]
+        else:
+            i = low + (high - low) * torch.rand(*size, **attrs)
+        return r + 1j * i
```

Comment:
What's the status for complex number support in pytorch? As far as I can tell, it's [work in progress](https://github.com/pytorch/pytorch/issues/755). For instance, when I do the following, I don't get an error, but also not what I would expect:
```
In [13]: a = torch.randn(1)

In [14]: a
Out[14]: tensor([2.2620])

In [15]: a * 1j
Out[15]: tensor([0.])

In [17]: a * 1j * 1j
Out[17]: tensor([0.])

# Expect: tensor([-1.])
```

Reply:
That's interesting..
When we remove the complex input support (as not supported atm) then code will simplify a bit. However, this might make restoring the complex input support difficult in the future as the algorithms need to be revisited again. 
So, I would recommend to leave it in as it is. What do you think?

Reply:
I think we should leave it in. Also cc'ing @dylanbespalko for a second opinion.

Reply:
> What's the status for complex number support in pytorch?

I have worked PyTorch. Here is the status:
- Complex number support is discussed here #755.
- Support is being added when kernels are ported from TH to ATen #24507. (Huge effort).

Since it is impossible to guarantee full support for complex numbers ATM, you can enable it with a PyTorch extension _(Note: PyTorch CI does not currently validate complex number support):_
- [cpu_strided_complex](https://gitlab.com/pytorch-complex/pytorch-cpu-strided-complex)
- [cuda_strided_complex](https://gitlab.com/pytorch-complex/pytorch-cuda-strided-complex)
- [fpga_strided_complex](https://gitlab.com/pytorch-complex/vitis_kernels) (work-in-progress)

Some notable kernels that are missing complex number support are ('matmul', 'mm', 'mv', 'rand'). I would recommend the following:
 - If you are worried about breaking complex number support by using a kernel that doesn't support complex numbers yet (eg. 'matmul'),  just do it and we'll fix complex number support later (PyTorch CI will not fail).
 - If you are taking advantage of an SVD simplification that is only valid for real numbers (eg. 'fft_real', 'norm2d'), please create a new kernel or provide separate plumbing for complex numbers.

Reply:
@dylanbespalko -- I read your statement as a green light on keeping the code relating to complex numbers. I'm ok with this.

Reply:
#33152 for reference